Added fixes to homeview, added DropDown and used device data in statu…

aleksikosonen opened the pull request:
fixes #18 

## Description

Added a `DropDown` component to select between the broadcasted Channels and Device statuses and filter the `DeviceList` based on it. Also enhanced layout, used devices `trustedState` as data in `StatusComponent` and refactored the DeviceList out of BaseView.

## Added components
- DropDownComponent
- DeviceList
- DropDownComponent.test.js

## Frontend
- The DetailsView needs to be though out again, as it should not re-render the `BaseView` components. That's why one StatusComponent is commented out with a comment regarding this. 
- Added `channelUrl` to globals
- Added `useChannels`to `ApiHooks` to get Channels from database
- Added `channelType` type
- Fixed DeviceListComponent layout to match available space

## Backend
- Updated Cors-policy to deviceController and channelController
- Changed JSON to json in `channelController`
- Changed naming in Schema files to have first letter in uppercase, as CI/CD pipeline linux is case sensitive

aleksikosonen commented:
Realized the new components are missing Cypress tests so converted PR to Draft and will add them.

aleksikosonen commented:
Agreed with @jaanikaukonen that the dropdown should be styled a bit more before merge. Will do that later this weekend

jaanikaukonen commented on front-end/src/components/DropDownComponent/DropDown.tsx:
```diff
@@ -0,0 +1,68 @@
+  const [isDropDownOpen, setIsDropDownOpen] = useState(false)
+  const [selectedElementIndex, setSelectedElementIndex] = useState(0)
+
+  const handleClick = () => {
+    if (isDropDownOpen == false) {
+      setIsDropDownOpen(true)
+    } else {
+      setIsDropDownOpen(false)
+    }
+  }
```
```suggestion
  const handleClick = () => {
      setIsDropDownOpen(!isDropDownOpen)
  }
```

aleksikosonen replied:
Fixed this in the new commit!

NikoLindborg commented on front-end/src/views/base/BaseView.tsx:
```diff
@@ -31,9 +33,13 @@ const BaseView: React.FC = () => {
-  const handleClick = (id: string) => {
-    console.log(`device id, ${id}`)
-  }
+  devices.forEach((device) => {
+    device.channels?.forEach((channel) => {
+      if (!listOfChannels.includes(channel)) {
+        listOfChannels.push(channel)
+      }
+    })
+  })
```
We should just make an endpoint in the backend that provides channels from databases subscribed channels list with the announcement channel filtered instead of this

aleksikosonen replied:
Done initially this way to reduce database requests, could later on be done that way also if needed

NikoLindborg replied:
1 database request is way smaller problem than having code that gets rerun for no reason on every re-render that the component gets. Uncontrolled code inside a component is not a good idea, or a habit.

aleksikosonen replied:
Fixed this in the new commit, added useChannels to ApiHooks

NikoLindborg reviewed (commented):
Aside from the 1 comment looks good. If your going to be working on this branch more anyways, could include that change aswell

aleksikosonen commented:
## Description

Added UI to DropDowns and polished some things on the way. Also added logic to have multiple dropdowns and now devices can be filtered by devices status and subscribed channels. 

# Additional changes

- Changed the trusted values from 0, 1 & 2 to 1, 2 & 3 throughout the code and MongoDB to ease filtering (0 caused issues in filtering)

NikoLindborg reviewed (changes requested):
Please do the changes requested above

jaanikaukonen commented on front-end/src/components/DropDownComponent/DropDown.tsx:
```diff
@@ -0,0 +1,80 @@
+        }
+        style={{width: dropDownContentWidth}}
+      >
+        {elements.map((e, i) => (
```
```suggestion
        {elements.map((element, index) => (
```
Less cryptic imo.

aleksikosonen replied:
fixed in the next commit

jaanikaukonen commented on front-end/src/views/details/DetailsView.tsx:
```diff
@@ -95,9 +95,9 @@ const DetailsView: React.FC<DetailsViewProps> = ({id}) => {
             device.sensors.map((sensor) => (
               <DetailComponent
-                key={device._id}
+                key={device._id as number}
```
Does this refer the object id in MongoDb? In mongo, _id is an object, but it can be refer as string. Definitely not a number since it contains characters.

aleksikosonen replied:
Fixed in the new commit

jaanikaukonen commented on front-end/src/components/DropDownComponent/DropDown.tsx:
```diff
@@ -0,0 +1,80 @@
+import './DropDown.css'
+
+interface DropDownProps {
+  elements: String[]
+  selectedElement: String
```
```suggestion
  elements: string[]
  selectedElement: string
```
String is JavaScript type for string,
string is TypeScript type for string.

aleksikosonen replied:
fixed in the next commit

jaanikaukonen reviewed (changes requested):
Update pr description. There are changes in the backend as well, like added cors policy and so on. It is also missing status dropdown, no mention that device card is now wider than before. Also since we found a bug from the dropdown (they overlap if both are open), submit a bug item and put the link to the issue in the desc as well.

Other than those and open comments, looks good!

NikoLindborg reviewed (approved):
Looks good to me now
